src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs

```diff
@@ -160,18 +160,18 @@ public static void MoveFile(string sourceFullPath, string destFullPath)
         {
             // The desired behavior for Move(source, dest) is to not overwrite the destination file
             // if it exists. Since rename(source, dest) will replace the file at 'dest' if it exists,
-            // link/unlink are used instead. However, if the source path and the dest path refer to
-            // the same file, then do a rename rather than a link and an unlink.  This is important
-            // for case-insensitive file systems (e.g. renaming a file in a way that just changes casing),
-            // so that we support changing the casing in the naming of the file. If this fails in any
-            // way (e.g. source file doesn't exist, dest file doesn't exist, rename fails, etc.), we
-            // just fall back to trying the link/unlink approach and generating any exceptional messages
-            // from there as necessary.
+            // link/unlink are used instead. Rename is more efficient than link/unlink on file systems
+            // where hard links are not supported (such as FAT). Therefore, given that source file exists,
+            // rename is used in 2 cases: when dest file does not exist or when source path and dest
+            // path refer to the same file (on the same device). This is important for case-insensitive
+            // file systems (e.g. renaming a file in a way that just changes casing), so that we support
+            // changing the casing in the naming of the file. If this fails in any way (e.g. source file
+            // doesn't exist, dest file doesn't exist, rename fails, etc.), we just fall back to trying the
+            // link/unlink approach and generating any exceptional messages from there as necessary.
             Interop.Sys.FileStatus sourceStat, destStat;
             if (Interop.Sys.LStat(sourceFullPath, out sourceStat) == 0 && // source file exists
-                Interop.Sys.LStat(destFullPath, out destStat) == 0 && // dest file exists
-                sourceStat.Dev == destStat.Dev && // source and dest are on the same device
-                sourceStat.Ino == destStat.Ino && // and source and dest are the same file on that device
+               (Interop.Sys.LStat(destFullPath, out destStat) != 0 || // dest file does not exist
+                    sourceStat.Ino == destStat.Ino) && // source and dest are the same file on that device
                 Interop.Sys.Rename(sourceFullPath, destFullPath) == 0) // try the rename
```

Comment:
It looks like you removed the same device check here - `sourceStat.Dev == destStat.Dev`.  Is this intentional?

Reply:
@baulig I asked Mitch to essentially cherry pick the corefx change as it would solve the problem VS is having.  I noticed the change and wanted you to guide Mitch so that we would not regress on Android.  So, to answer your question, yes it was intentional, but we probably need to put your checks back in.  What do you think?

Reply:
Right, looking into it further, I'm not sure if it was intentional to call rename for the scenario in which `sourceFullPath` and `destFullPath` both lead to files that exist, but exist on different devices, yet still happen to have the same Inode number (file serial number). 
It appears that the original case was to catch a simple change in file case, ("aa.txt" -> "AA.txt"), and it looks like the cherry-picked PR was intended to widen the scope in which `Rename` is called and lower expectations of delete events firing.

Looking at the test cases, it's not obvious to me whether or not there is a check for moving a file from one device to another. Locally adding that same device check and testing, the intended functionality of Rename events firing is maintained.

Reply:
You are checking whether both files have the same inode (`sourceStat.Ino == destStat.Ino`), but without checking for the same device (`sourceStat.Dev == destStat.Dev`).

The inode check alone is not enough because two different files on two different devices may have the same inode by pure chance.  And when that happens, that code is now unexpectedly behaving differently.  And since that will be such a rare occurrence, it will be a total nightmare to track down.

Please either put back the device check or explain in a comment why the above mentioned scenario cannot happen, as I don't see that from looking at the code.

Reply:
When files exist on different devices, their inode numbers are completely meaningless and you shouldn't even attempt to "compare" them.  An inode number is just a way of uniquely indexing a file in a particular file system - and it's therefore unique to that file system.  They're also generated mostly randomly (not in a cryptographic sense of course!) and really have no special meaning other than ensuring uniqueness.

I spent some more time looking at that code and I believe the reason we are comparing the inode numbers is to identity the case where one of the files is a link to the other one - or when both are links to the same third file.  Using `lstat()` is far more efficient than `readlink()`.

Reply:
Or in other words, if you ever see a piece of code that looks at an inode number without also looking at the device id - something is likely wrong.

Reply:
I agree with @baulig. I missed seeing this removal, and the omission seems like an accident/error in the original PR as it is not addressed.

We should probably just re-add the line. We should bring this up in dotnet/runtime too.

Reply:
Thanks for the :eyes:. I re-added the logic to check that the source and dest files have the same device ID and inode serial number.

Reply:
Looks good to me now, but let's wait for the CoreFX folks to also address it on their end in https://github.com/dotnet/runtime/issues/34330

Reply:
https://github.com/dotnet/runtime/pull/34343 has squashed and merged